### ISSUE TITLE
Stream meshing immediately to STL?

### ISSUE THREAD

**Issue by lf94** [libfive/libfive#535]
Is it possible to stream the mesh immediately to a file? It'd be nice to do this to be able to render massive models without needing lots of memory! I've been looking at it tonight and it *appears* it wouldn't be too difficult. For every branes and verts `push_back`, it could instead write to a file in STL format.... Like here https://github.com/libfive/libfive/blob/e7486b9b63c1f67509be02399d09b9411ea6c1c2/libfive/src/render/brep/dc/dc_mesher.cpp#L166

Edit: something like https://github.com/keean/File-Vector could be useful.

Edit: I guess this all boils down to one assumption:
https://github.com/libfive/libfive/blob/e7486b9b63c1f67509be02399d09b9411ea6c1c2/libfive/src/render/brep/mesh.cpp#L182

Are all of the vertices per brane / face already known as each face is "created"?

**Comment by mkeeter** [comment]
Kinda, but there are a few complexities:
- Meshing builds an **indexed** mesh, e.g. `v[0]` in the example above is a `uint32_t`, rather than a `vec3f`.  This allows us to use less memory during meshing, since vertices are only stored once.  However, an STL is **not** an indexed representation, so we'd have to unpack into individual vertices
- A `PerThreadBRep` exists – like the name implies – on a per-thread basis.  We have many of them being constructed simultaneously; if all of those threads were trying to write to the same file, we've need some kind of locking or arbitration.

In short, this wouldn't be particularly easy, and I'm not totally convinced of the benefit; my intuition is that the majority of memory is allocated during octree construction, and the meshing step doesn't allocate _much_ more.

**Comment by lf94** [comment]
> my intuition is that the majority of memory is allocated during octree construction, and the meshing step doesn't allocate much more.

Gotchya. Then there's no real way forward with this without basically writing the octree to a file, manipulating the file, etc... I'll close :)
